Login: Refactor show and hide

garrett opened the pull request:
Partially addresses #14828.

garrett commented:
TODO:
- [ ] Convert all the showing and hiding to show() and hide()

garrett commented:
I think I successfully changed everything over to hidden attribute showing and hiding with `show()` and `hide()` (and `hideToggle()`, which shows or hides based on a boolean) in JavaScript and `hidden` in the HTML in commit f8b0325.

Please review and test!

garrett commented:
This refactoring was needed for #14688, to rip out arbitrary CSS in JavaScript, so I could properly implement the UI from the designs.

mvollmer commented:
The "unsupported browser" state looks like this:

![image](https://user-images.githubusercontent.com/3228183/98517827-33c60480-2277-11eb-9ba1-d56a3be9139a.png)

Shouldn't there be a message saying "Your browser is unsupported" or similar?  I check with master.

mvollmer commented:
"OAuth logout" looks like this:

![image](https://user-images.githubusercontent.com/3228183/98519138-05e1bf80-2279-11eb-8218-37d5094bc56a.png)

I check master as well. In any case, this doesn't look "designed". It has probably regressed over time.

mvollmer commented:
Same for "Oauth failure":

![image](https://user-images.githubusercontent.com/3228183/98519297-40e3f300-2279-11eb-90c8-2d92aa2dcffa.png)

mvollmer commented:
> Shouldn't there be a message saying "Your browser is unsupported" or similar? I check with master.

Same on master -> #14884

mvollmer commented:
> "OAuth logout" looks like this:

On master:

![image](https://user-images.githubusercontent.com/3228183/98521730-7dfdb480-227c-11eb-8eda-1d10fcfac4c2.png)

I think, this happens because `#login-again` gets `display: block` in the old code when it is shown.

mvollmer commented on src/ws/login.js:
```diff
@@ -74,21 +74,56 @@
         if (oauth_redirect_to) {
             id("login-again").href = oauth_redirect_to;
-            id("login-again").style.display = "block";
+            show("#login-again");
```
`#login-again` is a `a` element, and giving it `display: block` had an effect. I think we should keep it `display: block` somehow, like with a explicit rule in login.css.

mvollmer replied:
See https://github.com/cockpit-project/cockpit/pull/14870#issuecomment-723864991 and https://github.com/cockpit-project/cockpit/pull/14870#issuecomment-723878719

garrett replied:
Yep. The display types should be in CSS, not JavaScript. I'll add it to the CSS.

mvollmer reviewed (changes requested):
Very nice!

mvollmer commented on src/ws/login.js:
```diff
@@ -506,26 +538,26 @@
+        show("#login");
+
+        hideToggle(["#user-group", "#password-group"], in_conversation);
+        hideToggle("#conservation-group", !in_conversation);
```
This should explain most of the test failures.

garrett commented:
Rebased on current master; I'll look into and solve what I can now. I'll need some help with non-password logins. (Even though I guess they're currently broken in the same way on master too?)

garrett commented:
> Shouldn't there be a message saying "Your browser is unsupported" or similar? I check with master.

Yes. It's:

```
          <h2 class="unsupported-browser-heading" translate>A modern browser is required for security, reliability, and performance.</h3>
```
(Oh, opens with h2, closed with h3?! I'll fix that too.)

Is this rewritten in the JavaScript perhaps?

garrett commented:
Rebased, addressed concerns.

Specifically did not fix the browser message, as that's apparently broken on master as well. We should fix it, but probably in a follow-up, unless it's related to show/hide. (Which I don't think it is?)

mvollmer commented:
> > Shouldn't there be a message saying "Your browser is unsupported" or similar? I check with master.
> 
> Yes. It's:
> 
> https://github.com/cockpit-project/cockpit/blob/4113c2e89f102cfc5f7291edd8676a675cf3adc4/src/ws/login.html#L109
> 
> Is this rewritten in the JavaScript perhaps?

No, it is moved off-screen by the CSS:

```
  .unsupported-browser-heading {
    font-size: 3rem;
    left: 50%;
    /* Pull the H2 half the width to the left to center */
    margin: 0 0 0 -32rem;
    position: absolute;
    text-align: center;
    top: -16rem;
    width: 64rem;
  }
```

Let's continue in  #14884.

garrett commented:
> No, it is moved off-screen by the CSS

Oof. I wonder how that happened. That's definitely a hack. I guess that's from before I added grid.

garrett commented:
I've made a draft @ #14891 to fix no JavaScript and the browser support messaging. (The support messaging fixups aren't there yet.)

But the important thing is that we're considering this PR done, as those aren't regressions compared to what's currently in our main branch — and these issues (no JS, unsupported browser) will be addressed in the follow-up.
